src/app-layer-frames.c

```diff
@@ -0,0 +1,416 @@
+    }
+
+    const uint32_t slide = progress - STREAM_APP_PROGRESS(stream);
+    frames->progress_rel += slide;
+}
+
+void AppLayerFramesSlide(Flow *f, const uint32_t slide, const uint8_t diframetion)
+{
+    FramesContainer *frames_container = AppLayerFramesGetContainer(f);
+    if (frames_container == NULL)
+        return;
+    Frames *frames;
+    if (diframetion == STREAM_TOSERVER) {
+        frames = &frames_container->toserver;
+    } else {
+        frames = &frames_container->toclient;
+    }
+    FrameSlide(frames, slide);
+}
+
+static void FrameFreeSingleFrame(Frames *frames, Frame *r)
+{
+    FrameDebug("free", frames, r);
+    FrameClean(r);
+}
+
+void FramesFree(Frames *frames)
+{
+    BUG_ON(frames == NULL);
+
+    for (uint16_t i = 0; i < frames->cnt; i++) {
+        if (i < FRAMES_STATIC_CNT) {
+            Frame *r = &frames->sframes[i];
+            FrameFreeSingleFrame(frames, r);
+        } else {
+            const uint16_t o = i - FRAMES_STATIC_CNT;
+            Frame *r = &frames->dframes[o];
+            FrameFreeSingleFrame(frames, r);
+        }
+    }
+    SCFree(frames->dframes);
+    frames->dframes = NULL;
+}
+
+Frame *AppLayerFrameNew(Flow *f, const StreamSlice *stream_slice, const uint8_t *frame_start,
+        const uint32_t len, int dir, uint8_t frame_type)
+{
+    SCLogDebug("stream_slice offset %" PRIu64, stream_slice->offset);
+    SCLogDebug("frame_start %p stream_slice->input %p", frame_start, stream_slice->input);
+
+    /* workarounds for many (unit|fuzz)tests not handling TCP data properly */
+#if defined(UNITTESTS) || defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION)
+    if (f->protoctx == NULL)
+        return NULL;
+    if (frame_start < stream_slice->input || frame_start >= stream_slice->input + stream_slice->input_len)
+        return NULL;
+#endif
+    BUG_ON(frame_start < stream_slice->input);
+    BUG_ON(stream_slice->input == NULL);
+    BUG_ON(f->proto != IPPROTO_TCP);
+    BUG_ON(f->protoctx == NULL);
+
+    ptrdiff_t ptr_offset = frame_start - stream_slice->input;
+#ifdef DEBUG
+    uint64_t offset = ptr_offset + stream_slice->offset;
+    SCLogDebug("flow %p diframetion %s frame %p starting at %" PRIu64 " len %u (offset %" PRIu64
+               ")",
+            f, dir == 0 ? "toserver" : "toclient", frame_start, offset, len, stream_slice->offset);
+#endif
+    BUG_ON(f->alparser == NULL);
+
+    FramesContainer *frames_container = AppLayerFramesSetupContainer(f);
+    if (frames_container == NULL)
+        return NULL;
+
+    Frames *frames;
+    if (dir == 0) {
+        frames = &frames_container->toserver;
+    } else {
+        frames = &frames_container->toclient;
+    }
+
+    Frame *r = FrameNew(frames, (uint32_t)ptr_offset, len);
+    if (r != NULL) {
+        r->type = frame_type;
+    }
+    return r;
+}
+
+Frame *AppLayerFrameNew2(Flow *f, const StreamSlice *stream_slice, const uint32_t frame_start_rel,
+        const uint32_t len, int dir, uint8_t frame_type)
```

Comment:
Will there just be one variation when done?

Reply:
Hope so

Reply:
Kept 2 in #6684, but gave them clearer names